datacube/api/core.py

```diff
@@ -350,14 +348,32 @@ def group_datasets(datasets, group_by):
 
         .. seealso:: :meth:`find_datasets`, :meth:`load_data`, :meth:`query_group_by`
         """
+        if isinstance(group_by, str):
+            group_by = query_group_by(group_by=group_by)
+
         dimension, group_func, units, sort_key = group_by
-        datasets.sort(key=sort_key)
-        groups = [Group(key, tuple(group)) for key, group in groupby(datasets, group_func)]
 
-        data = numpy.empty(len(groups), dtype=object)
-        for index, group in enumerate(groups):
-            data[index] = group.datasets
-        coords = [sort_key(v.datasets[0]) for v in groups]
+        def ds_sorter(ds):
+            return sort_key(ds), getattr(ds, 'id', 0)
+
+        def mk_group(group):
+            dss = tuple(sorted(group, key=ds_sorter))
+            # TODO: decouple axis_value from group sorted order
+            axis_value = sort_key(dss[0])
+            return (axis_value, dss)
+
+        datasets = sorted(datasets, key=group_func)
```

Comment:
What about renaming `group_func` to `group_key` here as you proposed earlier?

Reply:
Yes we need better names. I think this should be done by replacing entire `GroupBy` tuple with something else. Something that is responsible for supplying policies for how to:

1. Group datasets
2. Sort datasets with group
3. Compute axis value for a group
4. Supply axis name

I suggested to change names in your PR because you were changing entire `GroupBy` object, here I'm leaving this interface as is for now.

Reply:
In the current implementation 2 and 3 are coupled, before this PR 1 and 2 were also coupled